mungegithub/mungers/publish_scripts/util.sh

```diff
@@ -19,110 +19,446 @@
+        local dst_branch_point_commit=$(branch-commit ${k_branch_point_commit} master)
+        if [ -z "${dst_branch_point_commit}" ]; then
+            # fall-back to the old "sync(k8s.io/kubernetes) <sha1>" commits
+            dst_branch_point_commit=$(git log --oneline --grep "sync(k8s.io/kubernetes) *${k_branch_point_commit}" --format='%H' master)
+            if [ -z "${dst_branch_point_commit}" ]; then
+                echo "Couldn't find a corresponding branch point commit for ${k_branch_point_commit} as ascendent of origin/master."
+                return 1
+            fi
+        fi
+
+        git branch -f ${dst_branch} ${dst_branch_point_commit} >/dev/null
+
+        # there might be merges from master into ${dst_branch}. We pick the last one as the base because everything after is linear again.
+        local filtered_branch_base_merge_commit=""
+        local parent1=""
+        local parent2=""
+        local filtered_branch_base_commit=$(git rev-parse filtered-branch-base)
+        read filtered_branch_base_merge_commit parent1 parent2 <<<$(git log --first-parent --format='%h %P' filtered-branch | grep ${filtered_branch_base_commit})
+        if [ -n "${filtered_branch_base_merge_commit}" ] && [ -n "${parent2}" ]; then
+            # only consider cherry-picks from here upwards
+            f_base_commit=${filtered_branch_base_merge_commit}
+            # remember the merge to be cherry-picked before all other mainline commits
+            f_base_merge_commit=${filtered_branch_base_merge_commit}
+
+            # find which parent is the filtered-branch-base, necessary for cherry-picking this merge
+            if [ "${parent1}" = ${filtered_branch_base_commit} ]; then
+                f_base_merge_parent=1
+            elif [ "${parent2}" = ${filtered_branch_base_commit} ]; then
+                f_base_merge_parent=2
+            else
+                echo "Expected the branch point ${filtered_branch_base_commit} to be either first or second parent of ${filtered_branch_base_merge_commit}."
+                return 1
+            fi
+        else
+            # only consider cherry-picks from the branch point upwards
+            f_base_commit=${filtered_branch_base_commit}
+        fi
+    elif [ "${new_repo}" = false ]; then
+        # old repo, use filtered-branch-base as basis for further cherry-picking
+        f_base_commit=$(git rev-parse filtered-branch-base)
+    fi
+
+    # find commits on the main line (mostly merges, some commits due to fast-forward merges that filter-branch dropped)
+    local f_mainline_commits=""
+    if [ -z "${f_base_commit}" ]; then
+        f_mainline_commits=$(git log --first-parent --format='%H' --topo-order --reverse HEAD)
+    else
+        f_mainline_commits=$(git log --first-parent --format='%H' --topo-order --reverse ${f_base_commit}..HEAD)
```

Comment:
A silly question, doesn't the  `--first-parent` make output commits linear? If so, why do we still need --topo-order?

Reply:
Good point. topo-order can go.

Reply:
fixed